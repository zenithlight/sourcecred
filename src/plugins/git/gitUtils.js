// @flow

import {execFileSync} from "child_process";
import fs from "fs-extra";
import path from "path";

export interface Utils {
  exec: GitDriver;
  head(): string;
  writeAndStage(filename: string, contents: string): void;
  deterministicCommit(message: string): void;
}

export type GitDriver = (args: string[], env?: {[string]: string}) => string;

export function localGit(repositoryPath: string): GitDriver {
  return function git(args: string[], env?: {[string]: string}): string {
    // We standardize the environment variables shown to Git, using
    // Git's test suite [1] as inspiration. It is particularly important
    // that `GIT_DIR` be unset from the parent process environment.
    // Otherwise, these tests have the wrong behavior when running in an
    // `exec` step of a Git rebase.
    //
    // [1]: https://github.com/git/git/blob/1f1cddd558b54bb0ce19c8ace353fd07b758510d/t/test-lib.sh#L90
    const baseEnv: {[string]: string | void} = {
      // Standardize output.
      LANG: "C",
      LC_ALL: "C",
      PAGER: "cat",
      TZ: "UTC",
      // Short-circuit editing.
      EDITOR: "true", // (this is `true` the command-line program)
      GIT_MERGE_AUTOEDIT: "no",
      // Ignore global Git settings, for test isolation.
      GIT_CONFIG_NOSYSTEM: "1",
      GIT_ATTR_NOSYSTEM: "1",
      // Bring over the SSH configuration so that loading private repos is possible
      // This post has some useful information on SSH_AUTH_SOCK:
      // http://blog.joncairns.com/2013/12/understanding-ssh-agent-and-ssh-add/
      SSH_AUTH_SOCK: process.env.SSH_AUTH_SOCK,
    };
    const fullEnv: {[string]: string | void} = {
      ...baseEnv,
      ...(env || {}: {+[string]: string | void}),
    };
    const options = {env: fullEnv};
    return execFileSync(
      "git",
      ["-C", repositoryPath, ...args],
      options
    ).toString();
  };
}

export function makeUtils(repositoryPath: string): Utils {
  const git = localGit(repositoryPath);
  return {
    exec: git,

    head() {
      return git(["rev-parse", "HEAD"]).trim();
    },

    writeAndStage(filename: string, contents: string) {
      const filepath = path.join(repositoryPath, filename);
      const dirpath = path.join(repositoryPath, path.dirname(filename));
      fs.mkdirpSync(dirpath);
      fs.writeFileSync(filepath, contents);
      git(["add", filename]);
    },

    deterministicCommit(message: string): void {
      git(
        [
          "-c",
          "user.name=Test Runner",
          "-c",
          "user.email=nobody@example.com",
          "commit",
          "-m",
          message,
        ],
        {
          GIT_AUTHOR_DATE: "2001-02-03T04:05:06",
          GIT_COMMITTER_DATE: "2002-03-04T05:06:07",
        }
      );
    },
  };
}
