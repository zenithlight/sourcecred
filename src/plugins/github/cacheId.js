// @flow

import {type RepoId, repoIdToString} from "./repoId";

/**
 * A derived ID to reference a cache layer.
 */
export opaque type CacheId: string = string;

/**
 * Derives the CacheId for a RepoId.
 *
 * Returned `CacheId`s will be:
 * - Deterministic
 * - Unique for this plugin
 * - Lowercase
 * - Safe to use for filenames
 * - Distinct for semantically distinct inputs (input IDs that differ
 *   only in case may map to the same output, because GitHub does not
 *   permit collisions-modulo-case)
 */
export function cacheIdForRepoId(repoId: RepoId): CacheId {
  // GitHub owner (user/organization) and repository names may be in
  // mixed case, but GitHub prevents mixed-case collisions: e.g., if
  // `foo` is a user then `Foo` cannot also be a user, and likewise for
  // repositories. GitHub login names are DNS-safe (`[0-9A-Za-z-]` only)
  // and so are filename-safe. Repository names have a slightly larger
  // character set, including underscore, but are also filename-safe.
  // Because login may not contain an underscore, it thus suffices to
  // use an underscore as the delimiter. The resulting filename uses
  // only `[0-9A-Za-z_.-]` and so is valid on all major filesystems.
  const owner = repoId.owner.toLowerCase();
  const name = repoId.name.toLowerCase();
  if (owner.includes("_")) {
    throw new Error(
      "unexpected underscore in GitHub owner name would be ambiguous: " +
        repoIdToString(repoId)
    );
  }
  return `github_${owner}_${name}`;
}
