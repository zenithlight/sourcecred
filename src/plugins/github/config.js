// @flow

import * as Combo from "../../util/combo";
import {type RepoId, stringToRepoId} from "./repoId";

export type GithubConfig = {|
  +repoIds: $ReadOnlyArray<RepoId>,
|};

// eslint-disable-next-line no-unused-vars
type SerializedGithubConfig = {|
  +repositories: $ReadOnlyArray<string>,
|};
// (^ for documentation purposes)

type JsonObject =
  | string
  | number
  | boolean
  | null
  | JsonObject[]
  | {[string]: JsonObject};

export const parser: Combo.Parser<GithubConfig> = (() => {
  const C = Combo;
  return C.object({
    repoIds: C.rename(
      "repositories",
      C.array(C.fmap(C.string, stringToRepoId))
    ),
  });
})();
