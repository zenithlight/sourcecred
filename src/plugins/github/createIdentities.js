// @flow

import {type IdentityProposal} from "../../core/ledger/identityProposal";
import {coerce, nameFromString} from "../../core/identity/name";
import {RelationalView, Userlike} from "./relationalView";
import {toRaw} from "./nodes";

export function _createIdentity(u: Userlike): IdentityProposal {
  const alias = {
    description: u.description(),
    address: toRaw(u.address()),
  };
  function chooseType(u: Userlike) {
    const subtype = u.address().subtype;
    // TODO: We should let it infer ORGANIZATION or PROJECT types too,
    // but at present we'll let the maintainer set this directly in the UI.
    switch (subtype) {
      case "USER": {
        return "USER";
      }
      case "BOT": {
        return "BOT";
      }
      default:
        throw new Error(`unknown userlike subtype: ${(subtype: empty)}`);
    }
  }
  return {
    pluginName: nameFromString("github"),
    name: coerce(u.login()),
    type: chooseType(u),
    alias,
  };
}

export function createIdentities(
  rv: RelationalView
): $ReadOnlyArray<IdentityProposal> {
  return Array.from(rv.userlikes()).map(_createIdentity);
}
