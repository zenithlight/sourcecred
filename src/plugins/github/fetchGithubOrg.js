// @flow

import {type RepoId, makeRepoId} from "./repoId";
import * as Queries from "../../graphql/queries";
import {postQuery} from "./fetchGithubRepo";
import {type GithubToken} from "./token";

export type Organization = {|
  +repos: $ReadOnlyArray<RepoId>,
  +name: string,
|};

const DEFAULT_PAGE_SIZE = 100;

/**
 * Fetches information about a given GitHub organization.
 *
 * Currently just gets the ids of its repositories, but we may want additional
 * information in the future.
 */
export async function fetchGithubOrg(
  org: string,
  token: GithubToken,
  // Regular clients should leave pageSize at the default 50.
  // Exposed for testing purposes.
  pageSize: ?number
): Promise<Organization> {
  const numRepos = pageSize == null ? DEFAULT_PAGE_SIZE : pageSize;
  const b = Queries.build;
  const makePayload = (afterCursor: ?string) => {
    const afterArg = afterCursor == null ? {} : {after: b.literal(afterCursor)};
    const args = {
      query: b.variable("searchQuery"),
      type: b.enumLiteral("REPOSITORY"),
      first: b.literal(numRepos),
      ...afterArg,
    };
    return {
      body: [
        b.query(
          "PerformSearch",
          [b.param("searchQuery", "String!")],
          [
            b.field("search", args, [
              b.field("nodes", {}, [
                b.inlineFragment("Repository", [
                  b.field("name"),
                  b.field("id"),
                ]),
              ]),
              b.field("pageInfo", {}, [
                b.field("endCursor"),
                b.field("hasNextPage"),
              ]),
            ]),
          ]
        ),
      ],
      variables: {searchQuery: `org:${org}`},
    };
  };

  let result = await postQuery(makePayload(), token);
  const resultNodes = [result.search.nodes];
  while (result.search.pageInfo.hasNextPage) {
    const afterCursor = result.search.pageInfo.endCursor;
    result = await postQuery(makePayload(afterCursor), token);
    resultNodes.push(result.search.nodes);
  }
  const repos = [].concat(...resultNodes).map((n) => makeRepoId(org, n.name));
  return {repos, name: org};
}
