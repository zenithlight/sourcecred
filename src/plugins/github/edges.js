// @flow

import deepFreeze from "deep-freeze";
import {
  type Edge,
  type EdgeAddressT,
  EdgeAddress,
  NodeAddress,
} from "../../core/graph";
import * as GithubNode from "./nodes";
import * as GitNode from "../git/nodes";
import {
  type ReactionContent,
  ReactionContent$Values as Reactions,
} from "./graphqlTypes";
import {type TimestampMs} from "../../util/timestamp";

export opaque type RawAddress: EdgeAddressT = EdgeAddressT;

export const AUTHORS_TYPE = "AUTHORS";
export const MERGED_AS_TYPE = "MERGED_AS";
export const HAS_PARENT_TYPE = "HAS_PARENT";
export const REFERENCES_TYPE = "REFERENCES";
export const REACTS_TYPE = "REACTS";
// GitHub tracks its own notion of a commit, which has a particular
// database id, is scoped to a particular repository, and has a canonical url
// on GitHub. The CORRESPONDS_TO_COMMIT_TYPE edges connect the GitHub commits
// to the corresponding Git commit.
export const CORRESPONDS_TO_COMMIT_TYPE = "CORRESPONDS_TO_COMMIT_TYPE";

const GITHUB_PREFIX = EdgeAddress.fromParts(["sourcecred", "github"]);
function githubEdgeAddress(...parts: string[]): RawAddress {
  return EdgeAddress.append(GITHUB_PREFIX, ...parts);
}

export const Prefix: {|
  authors: RawAddress,
  base: EdgeAddressT,
  correspondsToCommit: RawAddress,
  hasParent: RawAddress,
  mergedAs: RawAddress,
  reacts: RawAddress,
  reactsHeart: RawAddress,
  reactsHooray: RawAddress,
  reactsRocket: RawAddress,
  reactsThumbsUp: RawAddress,
  references: RawAddress,
|} = deepFreeze({
  base: GITHUB_PREFIX,
  authors: githubEdgeAddress(AUTHORS_TYPE),
  mergedAs: githubEdgeAddress(MERGED_AS_TYPE),
  references: githubEdgeAddress(REFERENCES_TYPE),
  hasParent: githubEdgeAddress(HAS_PARENT_TYPE),
  reacts: githubEdgeAddress(REACTS_TYPE),
  reactsThumbsUp: githubEdgeAddress(REACTS_TYPE, Reactions.THUMBS_UP),
  reactsHeart: githubEdgeAddress(REACTS_TYPE, Reactions.HEART),
  reactsHooray: githubEdgeAddress(REACTS_TYPE, Reactions.HOORAY),
  reactsRocket: githubEdgeAddress(REACTS_TYPE, Reactions.ROCKET),
  correspondsToCommit: githubEdgeAddress(CORRESPONDS_TO_COMMIT_TYPE),
});

export type AuthorsAddress = {|
  +type: typeof AUTHORS_TYPE,
  +author: GithubNode.UserlikeAddress,
  +content: GithubNode.AuthorableAddress,
|};
export type MergedAsAddress = {|
  +type: typeof MERGED_AS_TYPE,
  +pull: GithubNode.PullAddress,
|};
export type HasParentAddress = {|
  +type: typeof HAS_PARENT_TYPE,
  +child: GithubNode.ChildAddress,
|};
export type ReferencesAddress = {|
  +type: typeof REFERENCES_TYPE,
  +referrer: GithubNode.TextContentAddress,
  +referent: GithubNode.ReferentAddress,
|};
export type ReactsAddress = {|
  +type: typeof REACTS_TYPE,
  +reactionType: ReactionContent,
  +user: GithubNode.UserlikeAddress,
  +reactable: GithubNode.ReactableAddress,
|};
export type CorrespondsToCommitAddress = {|
  +type: typeof CORRESPONDS_TO_COMMIT_TYPE,
  +githubCommit: GithubNode.CommitAddress,
|};

export type StructuredAddress =
  | AuthorsAddress
  | MergedAsAddress
  | HasParentAddress
  | ReferencesAddress
  | ReactsAddress
  | CorrespondsToCommitAddress;

export const createEdge: {|
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ) => Edge,
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ) => Edge,
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ) => Edge,
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ) => Edge,
|} = deepFreeze({
  authors: (
    author: GithubNode.UserlikeAddress,
    content: GithubNode.AuthorableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: AUTHORS_TYPE, author, content}),
    src: GithubNode.toRaw(author),
    dst: GithubNode.toRaw(content),
    timestampMs,
  }),
  mergedAs: (
    pull: GithubNode.PullAddress,
    commit: GithubNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: MERGED_AS_TYPE, pull}),
    src: GithubNode.toRaw(pull),
    dst: GithubNode.toRaw(commit),
    timestampMs,
  }),
  correspondsToCommit: (
    githubCommit: GithubNode.CommitAddress,
    gitCommit: GitNode.CommitAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: CORRESPONDS_TO_COMMIT_TYPE, githubCommit}),
    src: GithubNode.toRaw(githubCommit),
    dst: GitNode.toRaw(gitCommit),
    timestampMs,
  }),
  hasParent: (
    child: GithubNode.ChildAddress,
    parent: GithubNode.ParentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: HAS_PARENT_TYPE, child}),
    src: GithubNode.toRaw(child),
    dst: GithubNode.toRaw(parent),
    timestampMs,
  }),
  references: (
    referrer: GithubNode.TextContentAddress,
    referent: GithubNode.ReferentAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({type: REFERENCES_TYPE, referrer, referent}),
    src: GithubNode.toRaw(referrer),
    dst: GithubNode.toRaw(referent),
    timestampMs,
  }),
  reacts: (
    reactionType: ReactionContent,
    user: GithubNode.UserlikeAddress,
    reactable: GithubNode.ReactableAddress,
    timestampMs: TimestampMs
  ): Edge => ({
    address: toRaw({
      type: REACTS_TYPE,
      user,
      reactionType,
      reactable,
    }),
    src: GithubNode.toRaw(user),
    dst: GithubNode.toRaw(reactable),
    timestampMs,
  }),
});

function lengthEncode(x: GithubNode.RawAddress): $ReadOnlyArray<string> {
  const parts = NodeAddress.toParts(x);
  return [String(parts.length), ...parts];
}
function lengthDecode(
  x: $ReadOnlyArray<string>,
  fail: () => Error
): {|+parts: $ReadOnlyArray<string>, +rest: $ReadOnlyArray<string>|} {
  if (x.length === 0) {
    // Not length-encoded.
    throw fail();
  }
  const [lengthString, ...allParts] = x;
  const length = parseInt(lengthString, 10);
  if (isNaN(length)) {
    throw fail();
  }
  if (length > allParts.length) {
    // Not enough elements.
    throw fail();
  }
  return {parts: allParts.slice(0, length), rest: allParts.slice(length)};
}
function multiLengthDecode(x: $ReadOnlyArray<string>, fail: () => Error) {
  let remaining = x;
  const partses = [];
  while (remaining.length > 0) {
    const {parts, rest} = lengthDecode(remaining, fail);
    partses.push(parts);
    remaining = rest;
  }
  return partses;
}

export function fromRaw(x: RawAddress): StructuredAddress {
  function fail() {
    return new Error(`Bad address: ${EdgeAddress.toString(x)}`);
  }
  if (!EdgeAddress.hasPrefix(x, GITHUB_PREFIX)) {
    throw fail();
  }
  const [_unused_sc, _unused_gh, kind, ...rest] = EdgeAddress.toParts(x);
  switch (kind) {
    case AUTHORS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [authorParts, contentParts] = parts;
      const author: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(authorParts): any)
      ): any);
      const content: GithubNode.AuthorableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(contentParts): any)
      ): any);
      return ({type: AUTHORS_TYPE, author, content}: AuthorsAddress);
    }
    case MERGED_AS_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [pullParts] = parts;
      const pull: GithubNode.PullAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(pullParts): any)
      ): any);
      return ({type: MERGED_AS_TYPE, pull}: MergedAsAddress);
    }
    case HAS_PARENT_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 1) {
        throw fail();
      }
      const [childParts] = parts;
      const child: GithubNode.ChildAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(childParts): any)
      ): any);
      return ({type: HAS_PARENT_TYPE, child}: HasParentAddress);
    }
    case REFERENCES_TYPE: {
      const parts = multiLengthDecode(rest, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [referrerParts, referentParts] = parts;
      const referrer: GithubNode.TextContentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referrerParts): any)
      ): any);
      const referent: GithubNode.ReferentAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(referentParts): any)
      ): any);
      return ({type: REFERENCES_TYPE, referrer, referent}: ReferencesAddress);
    }
    case REACTS_TYPE: {
      const [rawReactionType, ...rest2] = rest;
      const reactionType = Reactions[rawReactionType];
      if (reactionType == null) {
        throw fail();
      }
      const parts = multiLengthDecode(rest2, fail);
      if (parts.length !== 2) {
        throw fail();
      }
      const [userParts, reactableParts] = parts;
      const user: GithubNode.UserlikeAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(userParts): any)
      ): any);
      const reactable: GithubNode.ReactableAddress = (GithubNode.fromRaw(
        (NodeAddress.fromParts(reactableParts): any)
      ): any);
      return {type: REACTS_TYPE, reactionType, user, reactable};
    }
    default:
      throw fail();
  }
}

export function toRaw(x: StructuredAddress): RawAddress {
  switch (x.type) {
    case AUTHORS_TYPE:
      return EdgeAddress.append(
        Prefix.authors,
        ...lengthEncode(GithubNode.toRaw(x.author)),
        ...lengthEncode(GithubNode.toRaw(x.content))
      );
    case MERGED_AS_TYPE:
      return EdgeAddress.append(
        Prefix.mergedAs,
        ...lengthEncode(GithubNode.toRaw(x.pull))
      );
    case HAS_PARENT_TYPE:
      return EdgeAddress.append(
        Prefix.hasParent,
        ...lengthEncode(GithubNode.toRaw(x.child))
      );
    case REFERENCES_TYPE:
      return EdgeAddress.append(
        Prefix.references,
        ...lengthEncode(GithubNode.toRaw(x.referrer)),
        ...lengthEncode(GithubNode.toRaw(x.referent))
      );
    case REACTS_TYPE:
      return EdgeAddress.append(
        Prefix.reacts,
        x.reactionType,
        ...lengthEncode(GithubNode.toRaw(x.user)),
        ...lengthEncode(GithubNode.toRaw(x.reactable))
      );
    case CORRESPONDS_TO_COMMIT_TYPE:
      return EdgeAddress.append(
        Prefix.correspondsToCommit,
        ...lengthEncode(GithubNode.toRaw(x.githubCommit))
      );
    default:
      throw new Error((x.type: empty));
  }
}
