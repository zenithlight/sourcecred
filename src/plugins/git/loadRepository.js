/*
 * Load a git repository into memory. This dumps the commit and tree
 * data into a structured form. Contents of blobs are not loaded.
 *
 * If the repository contains file names that are not valid UTF-8
 * strings, the result is undefined.
 *
 * Note: git(1) is a runtime dependency of this module.
 */
// @flow

import * as MapUtil from "../../util/map";
import type {GitDriver} from "./gitUtils";
import type {Repository, Commit} from "./types";
import {localGit} from "./gitUtils";
import {repoIdToString, type RepoId, type RepoIdString} from "../github/repoId";

/**
 * Load a Git repository from disk into memory. The `rootRef` should be
 * a revision reference as accepted by `git rev-parse`: "HEAD" and
 * "origin/master" will be common, while a specific SHA or tag might be
 * used to fix a particular state of a repository.
 */
export function loadRepository(
  repositoryPath: string,
  rootRef: string,
  repoId: RepoId
): Repository {
  const git = localGit(repositoryPath);
  try {
    // If the repository is empty, HEAD will not exist. We check HEAD
    // rather than the provided `rootRef` because, in the case where the
    // repository is non-empty but the provided `rootRef` does not
    // exist, we still want to fail.
    git(["rev-parse", "--verify", "HEAD"]);
  } catch (e) {
    // No data in the repository.
    return {commits: {}, commitToRepoId: {}};
  }
  const rawCommits = findCommits(git, rootRef);
  const commits = MapUtil.toObject(new Map(rawCommits.map((x) => [x.hash, x])));
  const repoIdString = repoIdToString(repoId);
  const repoIdStringSet: () => {[RepoIdString]: true} = () => ({
    [((repoIdString: RepoIdString): any)]: true,
  });
  const commitToRepoId = MapUtil.toObject(
    new Map(rawCommits.map(({hash}) => [hash, repoIdStringSet()]))
  );
  return {commits, commitToRepoId};
}

function findCommits(git: GitDriver, rootRef: string): Commit[] {
  return git(["log", "--format=%H %h %at %P|%s", rootRef])
    .split("\n")
    .filter((line) => line.length > 0)
    .map((line) => {
      const pipeLocation = line.indexOf("|");
      const first = line.slice(0, pipeLocation).trim();
      const summary = line.slice(pipeLocation + 1);
      const [hash, shortHash, authorDate, ...parentHashes] = first.split(" ");
      const createdAt = +authorDate * 1000;
      return {hash, shortHash, createdAt, summary, parentHashes};
    });
}
