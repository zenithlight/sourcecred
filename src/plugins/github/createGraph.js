// @flow

import * as NullUtil from "../../util/null";
import {Graph} from "../../core/graph";
import {type WeightedGraph} from "../../core/weightedGraph";
import {type WeightsT, empty as emptyWeights} from "../../core/weights";
import * as GitNode from "../git/nodes";
import * as N from "./nodes";
import * as R from "./relationalView";
import {createEdge} from "./edges";
import {ReactionContent$Values as Reactions} from "./graphqlTypes";

export function createGraph(view: R.RelationalView): WeightedGraph {
  const creator = new GraphCreator();
  creator.addData(view);
  return {graph: creator.graph, weights: creator.weights};
}

class GraphCreator {
  graph: Graph;
  weights: WeightsT;

  constructor() {
    this.graph = new Graph();
    this.weights = emptyWeights();
  }

  addData(view: R.RelationalView) {
    for (const entity of view.entities()) {
      const address = N.toRaw(entity.address());
      this.graph.addNode({
        address,
        description: entity.description(),
        timestampMs: entity.timestampMs(),
      });
    }

    for (const child of view.childEntities()) {
      this.addHasParent(child);
    }

    for (const authored of view.authoredEntities()) {
      this.addAuthors(authored);
    }

    for (const pull of view.pulls()) {
      const commit = pull.mergedAs();
      if (commit != null) {
        const commitTimestamp = NullUtil.get(view.commit(commit)).timestampMs();
        this.graph.addEdge(
          createEdge.mergedAs(pull.address(), commit, commitTimestamp)
        );
      } else {
        const addr = N.toRaw(pull.address());
        // Un-merged PRs do not mint cred.
        this.weights.nodeWeights.set(addr, 0);
      }
    }

    for (const commit of view.commits()) {
      const gitCommitAddress: GitNode.CommitAddress = {
        type: GitNode.COMMIT_TYPE,
        hash: commit.hash(),
      };
      this.graph.addEdge(
        createEdge.correspondsToCommit(
          commit.address(),
          gitCommitAddress,
          commit.timestampMs()
        )
      );
    }

    for (const referrer of view.textContentEntities()) {
      for (const referent of referrer.references()) {
        this.graph.addEdge(
          createEdge.references(
            referrer.address(),
            referent.address(),
            referrer.timestampMs()
          )
        );
      }
    }

    for (const reactable of view.reactableEntities()) {
      for (const {content, user, timestampMs} of reactable.reactions()) {
        // We only support unambiguously positive reactions for now
        if (
          content === Reactions.THUMBS_UP ||
          content === Reactions.HEART ||
          content === Reactions.HOORAY ||
          content === Reactions.ROCKET
        ) {
          this.graph.addEdge(
            createEdge.reacts(content, user, reactable.address(), timestampMs)
          );
        }
      }
    }
  }

  addAuthors(entity: R.AuthoredEntity) {
    for (const author of entity.authors()) {
      this.graph.addEdge(
        createEdge.authors(
          author.address(),
          entity.address(),
          entity.timestampMs()
        )
      );
    }
  }

  addHasParent(child: R.Issue | R.Pull | R.Comment | R.Review) {
    this.graph.addEdge(
      createEdge.hasParent(
        child.address(),
        child.parent().address(),
        child.timestampMs()
      )
    );
  }
}
