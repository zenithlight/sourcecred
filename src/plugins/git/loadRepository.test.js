// @flow

import tmp from "tmp";

import {makeRepoId, repoIdToString} from "../github/repoId";
import {createExampleRepo} from "./example/exampleRepo";
import {localGit} from "./gitUtils";
import {loadRepository} from "./loadRepository";

const cleanups: (() => void)[] = [];
afterAll(() => {
  cleanups.forEach((f) => {
    f();
  });
});

function mkdtemp() {
  const result = tmp.dirSync({unsafeCleanup: true});
  cleanups.push(() => result.removeCallback());
  return result.name;
}

describe("plugins/git/loadRepository", () => {
  // Disabled while we dont have facilities for actually loading the git plugin
  it.skip("loads from HEAD", () => {
    const repository = createExampleRepo(mkdtemp());
    // In case of failure, run
    //     src/plugins/git/loadRepositoryTest.sh --updateSnapshot
    // to update the snapshot, then inspect the resulting changes.
    expect(
      loadRepository(
        repository.path,
        "HEAD",
        makeRepoId("sourcecred-test", "example-git")
      )
    ).toEqual(require("./example/example-git.json"));
  });

  it("sets the right repoId for every commit", () => {
    const gitRepository = createExampleRepo(mkdtemp());
    const repoId = makeRepoId("sourcecred-test", "example-git");
    const repository = loadRepository(gitRepository.path, "HEAD", repoId);
    for (const commitHash of Object.keys(repository.commits)) {
      expect(Object.keys(repository.commitToRepoId[commitHash])).toEqual([
        repoIdToString(repoId),
      ]);
    }
  });

  it("processes an old commit", () => {
    const repository = createExampleRepo(mkdtemp());
    const whole = loadRepository(
      repository.path,
      "HEAD",
      makeRepoId("sourcecred-test", "example-git")
    );
    const part = loadRepository(
      repository.path,
      repository.commits[1],
      makeRepoId("sourcecred-test", "example-git")
    );

    // Check that `part` is a subset of `whole`...
    Object.keys(part.commits).forEach((hash) => {
      expect(part.commits[hash]).toEqual(whole.commits[hash]);
    });

    // ...and that it's the right subset.
    expect({
      commits: new Set(Object.keys(part.commits)),
    }).toMatchSnapshot();
  });

  it("fails when given a non-existent root ref", () => {
    const repository = createExampleRepo(mkdtemp());
    const invalidHash = "0".repeat(40);
    expect(() => {
      loadRepository(
        repository.path,
        invalidHash,
        makeRepoId("sourcecred-test", "example-git")
      );
    }).toThrow("fatal: bad object 0000000000000000000000000000000000000000");
    expect.assertions(1);
  });

  it("handles an empty repository properly", () => {
    const repositoryPath = mkdtemp();
    const git = localGit(repositoryPath);
    git(["init"]);
    expect(
      loadRepository(
        repositoryPath,
        "HEAD",
        makeRepoId("sourcecred-test", "example-git")
      )
    ).toEqual({
      commits: {},
      commitToRepoId: {},
    });
  });
});
