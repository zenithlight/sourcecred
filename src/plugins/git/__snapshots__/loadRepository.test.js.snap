// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugins/git/loadRepository processes an old commit 1`] = `
Object {
  "commits": Set {
    "c08ee3a4edea384d5291ffcbf06724a13ed72325",
    "c2b51945e7457546912a8ce158ed9d294558d294",
  },
}
`;
